include/clang/AST/Expr.h

```diff
@@ -4514,6 +4514,120 @@ class GenericSelectionExpr : public Expr {
   friend class ASTStmtReader;
 };
 
+/// \brief Represents a Checked C bounds expression.
+class BoundsExpr : public Expr {
+private:
+  SourceLocation StartLoc, RParenLoc;
+
+public:
+  BoundsExpr(StmtClass StmtClass, SourceLocation StartLoc, SourceLocation RParenLoc)
+    : Expr(StmtClass, QualType(),  VK_RValue, OK_Ordinary, false,
+           false, false, false), StartLoc(StartLoc), RParenLoc(RParenLoc) {
+  }
+
+  explicit BoundsExpr(StmtClass StmtClass, EmptyShell Empty) :
+    Expr(StmtClass, Empty) {}
+
+  SourceLocation getStartLoc() { return StartLoc; }
+  void setStartLoc(SourceLocation Loc) { StartLoc = Loc; }
+  SourceLocation getRParenLoc() { return RParenLoc; }
+  void setRParenLoc(SourceLocation Loc) { RParenLoc = Loc; }
+
+  SourceLocation getLocStart() const LLVM_READONLY { return StartLoc; }
+  SourceLocation getLocEnd() const LLVM_READONLY { return RParenLoc; }
+};
+
+/// \brief Represents a Checked C nullary bounds expression.
+class NullaryBoundsExpr : public BoundsExpr {
+public:
+  enum Kind {
+    // Invalid bounds expressions are used to flag invalid bounds
+    // expressions and prevent spurious downstream error messages
+    // in bounds declaration checking.
+    Invalid = 0,
+    // bounds(none)
+    None = 1
+  };
+
+public:
+  NullaryBoundsExpr(Kind Kind, SourceLocation StartLoc, SourceLocation RParenLoc)
+    : BoundsExpr(NullaryBoundsExprClass, StartLoc, RParenLoc)  {
+    NullaryBoundsExprBits.Kind = Kind;
+  }
+
+  explicit NullaryBoundsExpr(EmptyShell Empty)
+    : BoundsExpr(NullaryBoundsExprClass, Empty) {}
+
+  Kind getKind() const { return (Kind) NullaryBoundsExprBits.Kind; }
+  void setKind(Kind Kind) { NullaryBoundsExprBits.Kind = Kind; }
+
+  // Iterators
+  child_range children() {
+    return child_range(child_iterator(), child_iterator());
+  }
+};
+
+/// \brief Represents a Checked C count bounds expression.
+class CountBoundsExpr : public BoundsExpr {
+public:
+  enum Kind {
+    Count = 0,
+    Byte_Count = 1,
+  };
+private:
+  Stmt* CountExpr;
+
+public:
+  CountBoundsExpr(Kind Kind, Expr* Count, SourceLocation StartLoc,
+    SourceLocation RParenLoc)
+    : BoundsExpr(CountBoundsExprClass, StartLoc, RParenLoc),
+      CountExpr(Count) {
+    CountBoundsExprBits.Kind = Kind;
+  }
+
+  explicit CountBoundsExpr(EmptyShell Empty)
+    : BoundsExpr(CountBoundsExprClass, Empty) {}
+
+  Kind getKind() const { return (Kind)CountBoundsExprBits.Kind; }
+  void setKind(Kind Kind) { CountBoundsExprBits.Kind = Kind; }
+
+  Expr *getCountExpr() const { return cast<Expr>(CountExpr); }
+  void setCountExpr(Expr *E) { CountExpr = E; }
+
+  // Iterators
+  child_range children() {
+    return child_range(&CountExpr, &CountExpr + 1);
```

Comment:
Other classes that have singleton children aren't using the SubExpr pattern, so that is why I used an explicit child here.

Note that with that pattern, I don't think we're getting a guarantee of a NULL Stmt pointer after the last valid expression.  If we have enum {A, B, LAST_ELEM}, LAST_ELEM will typically have the value 2, giving us a 2 element array sufficient to hold A & B.